src/content/developers/docs/smart-contracts/testing/index.md

```diff
@@ -1,37 +1,180 @@
+
+Manual-testing your smart contracts can also reveal vulnerabilities that exist outside the code, but can still affect it. For example, a smart contract audit can discover vulnerabilities arising from flawed interaction with off-chain components. 
+
+## Why is it important to test smart contracts? {#benefits-of-smart-contract-testing}
+
+Testing smart contracts is important for the following reasons:
+
+### 1. Smart contracts are high-value applications {#smart-contracts-are-high-value-applications}
+
+Smart contracts often deal with high-value financial assets, especially in industries like [decentralized finance (DeFi)](/defi/), and valuable items, such as [non-fungible tokens (NFTs)](/nft/). As such, minor vulnerabilities in smart contracts can and often lead to massive, irrecoverable losses for users. Comprehensive testing can, however, expose errors in smart contract code and reduce security risks before deployment. 
+
+### 2. Smart contracts are immutable {#smart-contracts-are-immutable}
+
+Smart contracts deployed in the [Ethereum Virtual Machine (EVM)](/developers/docs/evm/) are immutable by default. While traditional developers may be used to fixing software bugs after launching, Ethereum development leaves little room for patching security flaws once a smart contract is live on the blockchain. 
+
+While upgradeability mechanisms for smart contracts, such as proxy patterns, these can be difficult to implement. Besides reducing immutability and introducing complexity, upgrades often demand complex governance processes. 
+
+For the most part, upgrades should be considered a last resort and avoided unless necessary.  Detecting potential vulnerabilities and flaws in your smart contract during the pre-launch phase reduces the need for a logic upgrade. 
+
+## Automated testing for smart contracts {#automated-testing-for-smart-contracts}
+
+### 1. Unit testing {#unit-testing}
+
+Unit testing is an approach that involves testing individual components in a smart contract for correctness. A unit test is simple, quick to run, and provides a clear idea of what went wrong if the test fails. 
+
+Unit tests are crucial for smart contract development, especially if you need to add new logic to the code. Unit testing verifies the behavior of specific parts of your code and confirms that they execute as intended. 
+
+### 2. Static/dynamic analysis {#static-dynamic-analysis}
+
+Static analysis and dynamic analysis are two automated testing methods for evaluating the security qualities of smart contracts. Both techniques, however, use different approaches for finding defects in contract code. 
+
+- **Static analysis**
+
+Static analysis examines the source code or bytecode of a smart contract before execution. This means you can debug contract code without actually running the program. Static analyzers can detect common vulnerabilities in Ethereum smart contracts and aid compliance with best practices. 
+
+- **Dynamic analysis**
+
```

Comment:
Can we add some info about fuzzing too?

Reply:
Yup. I added a description for fuzzing now.